active/0016-emqx-conf.md

```diff
@@ -0,0 +1,184 @@
+- Unlike the previous declaration of all configuration items displayed in `emqx.conf`, if an item uses a default value, it does not need to be shown in `emqx.conf`.  the default is only embedded in the code. Also, users can view all configuration items via the HTTP API (described later). This would have 2 benefits: 
+  - In subsequent version upgrades, adding/removing/updating configurations will not be overwritten by `emqx.conf`. 
+  - This allows the user to focus only on the configuration that has been modified. 
+
+#### emqx_cluster_override.conf
+
+- This file can only be modified via the API and manual modification of file directly is not supported.
+
+- When updating configurations that must require consistency across the cluster, they are persisted to this file.
+- The node will copy this file from the longest surviving core node before initializing the configuration, , this file will be added to initialize the configuration together.
+- When the node is updated, the configuration within the cluster is updated via cluster call and persisted to this file.
+- This file must be kept the same for all nodes, we will add an extra process to check the content of the file periodically, and alarm after 3 continuous differences are found.
+
+#### emqx_local_override.conf
+
+- This file can only be modified via the API and manual modification of file directly is not supported.
+
+- When the configuration of a specific node is updated via HTTP API, it will be persisted to this file.
+
+#### emqx_conf application
+Before, the initialization of the configuration file, cluster_call, was done through the `emqx_machine`, we will split this part of the functionality from the` emqx_machine` and make a new application `emqx_conf`
+
+The role of this application is to
+- Converting the configuration from hocon format to erlang sys.config format at initialization.
+- Manage hot updates and deletions of configurations within the cluster.
+
+The basic node configuration hot update remains in emqx, in other words: `emqx_config_handler`, remains unchanged, but other apps that want to update the configuration must call through the `emqx_conf`'s' API, which cannot call emqx API directly.
+The specific flow is: 
+
+```bash
+Other Apps(eg: emqx_resource) => emqx_conf => emqx API.
+```
+
+### HTTP API design
+
+#### Get the whole configurations.
+
+```erlang
+#{
+  get => #{
+    description => <<"Get all the configurations of cluster, including hot and non-hot updatable items.">>,
+    parameters => [
+       {node, hoconsc:mk(typerefl:atom(),
+          #{in => query, required => false, example => <<"emqx@127.0.0.1">>,
+          desc => <<"Node's name: If you have this field defaulted, the whole cluster is return.">>})},
+       {debug, hoconsc:mk(typerefl:boolean(), #{in => query, required => false,
+          desc => <<"Carries debug(meta data) information, such as:line number,default value,documentation">>})}],
```

Comment:
I am not sure how easily fill in default value and documentation into metadata.
Maybe leave this feature out for now.

```suggestion
          desc => <<"Carries debug (metadata) information, such as file name and line number">>})}],
```

Reply:
ok, file name and line number first come.